express-notes-api/server.ts

```diff
@@ -0,0 +1,139 @@
+import express from 'express';
+import { readFile, writeFile } from 'node:fs/promises';
+import { ClientError } from './client-error.js';
+import { errorMiddleware } from './error-middleware.js';
+
+type Note = {
+  id: number;
+  content: string;
+};
+
+type Data = {
+  nextId: number;
+  notes: Record<number, Note>;
+};
+
+const app = express();
+const port = 8080;
+app.use(express.json());
+
+async function readData(): Promise<Data> {
+  const data = await readFile('./data.json', 'utf-8');
+  const dataParsed = JSON.parse(data);
+  return dataParsed;
+}
+
+async function writeData(data: Data): Promise<void> {
+  const entry = JSON.stringify(data, null, 2);
+  await writeFile('./data.json', entry, 'utf-8');
+}
+
+// GET a list of notes
+
+app.get('/api/notes', async (req, res, next) => {
+  try {
+    const notesArray: Note[] = [];
+    const data: Data = await readData();
+    for (const key in data.notes) {
+      notesArray.push(data.notes[key]);
+    }
+    res.status(200).json(notesArray);
+  } catch (error) {
+    next(error);
+  }
+});
+
+// GET a single note by ID
+
+app.get('/api/notes/:id', async (req, res, next) => {
+  try {
+    const id = Number(req.params.id);
+    const noteData = await readData();
+    const noteContent = noteData.notes[id];
+    if (id > 1) {
+      res.status(200).json(noteContent);
+      return;
+    }
+    if (id < 1 || Number.isNaN(id) || !Number.isInteger(id)) {
```

Comment:
I think the code is easier to read and work with if you do the error checking first, instead of last.